File: clients/scala/src/main/scala/ai_cup_22/codegame/ServerMessage.scala
package ai_cup_22.codegame

import ai_cup_22.util.StreamUtil

/**
 * Message sent from server
 */
sealed trait ServerMessage {
    /**
     * Write ServerMessage to output stream
     */
    def writeTo(stream: java.io.OutputStream): scala.Unit
}

object ServerMessage {
    /**
     * Update constants
     *
     * @param constants New constants
     */
    case class UpdateConstants(constants: ai_cup_22.model.Constants) extends ServerMessage {
        /**
         * Write UpdateConstants to output stream
         */
        override def writeTo(stream: java.io.OutputStream): scala.Unit = {
            StreamUtil.writeInt(stream, UpdateConstants.TAG)
            constants.writeTo(stream)
        }
    
        /**
         * Get string representation of UpdateConstants
         */
        override def toString(): String = {
            var stringBuilder = new StringBuilder("UpdateConstants { ")
            stringBuilder.append("constants: ")
            stringBuilder.append(constants)
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object UpdateConstants {
        val TAG: Int = 0
    
        /**
         * Read UpdateConstants from input stream
         */
        def readFrom(stream: java.io.InputStream): UpdateConstants = UpdateConstants(
            ai_cup_22.model.Constants.readFrom(stream)
        )
    }

    /**
     * Get order for next tick
     *
     * @param playerView Player's view
     * @param debugAvailable Whether app is running with debug interface available
     */
    case class GetOrder(playerView: ai_cup_22.model.Game, debugAvailable: Boolean) extends ServerMessage {
        /**
         * Write GetOrder to output stream
         */
        override def writeTo(stream: java.io.OutputStream): scala.Unit = {
            StreamUtil.writeInt(stream, GetOrder.TAG)
            playerView.writeTo(stream)
            StreamUtil.writeBoolean(stream, debugAvailable)
        }
    
        /**
         * Get string representation of GetOrder
         */
        override def toString(): String = {
            var stringBuilder = new StringBuilder("GetOrder { ")
            stringBuilder.append("playerView: ")
            stringBuilder.append(playerView)
            stringBuilder.append(", ")
            stringBuilder.append("debugAvailable: ")
            stringBuilder.append(debugAvailable)
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object GetOrder {
        val TAG: Int = 1
    
        /**
         * Read GetOrder from input stream
         */
        def readFrom(stream: java.io.InputStream): GetOrder = GetOrder(
            ai_cup_22.model.Game.readFrom(stream),
            StreamUtil.readBoolean(stream)
        )
    }

    /**
     * Signifies end of the game
     */
    case class Finish() extends ServerMessage {
        /**
         * Write Finish to output stream
         */
        override def writeTo(stream: java.io.OutputStream): scala.Unit = {
            StreamUtil.writeInt(stream, Finish.TAG)
        }
    
        /**
         * Get string representation of Finish
         */
        override def toString(): String = {
            var stringBuilder = new StringBuilder("Finish { ")
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object Finish {
        val TAG: Int = 2
    
        /**
         * Read Finish from input stream
         */
        def readFrom(stream: java.io.InputStream): Finish = Finish(
        )
    }

    /**
     * Debug update
     *
     * @param displayedTick Displayed tick
     */
    case class DebugUpdate(displayedTick: Int) extends ServerMessage {
        /**
         * Write DebugUpdate to output stream
         */
        override def writeTo(stream: java.io.OutputStream): scala.Unit = {
            StreamUtil.writeInt(stream, DebugUpdate.TAG)
            StreamUtil.writeInt(stream, displayedTick)
        }
    
        /**
         * Get string representation of DebugUpdate
         */
        override def toString(): String = {
            var stringBuilder = new StringBuilder("DebugUpdate { ")
            stringBuilder.append("displayedTick: ")
            stringBuilder.append(displayedTick)
            stringBuilder.append(" }")
            stringBuilder.toString()
        }
    }
    
    object DebugUpdate {
        val TAG: Int = 3
    
        /**
         * Read DebugUpdate from input stream
         */
        def readFrom(stream: java.io.InputStream): DebugUpdate = DebugUpdate(
            StreamUtil.readInt(stream)
        )
    }

    /**
     * Read ServerMessage from input stream
     */
    def readFrom(stream: java.io.InputStream): ServerMessage = {
        StreamUtil.readInt(stream) match {
            case UpdateConstants.TAG => UpdateConstants.readFrom(stream)
            case GetOrder.TAG => GetOrder.readFrom(stream)
            case Finish.TAG => Finish.readFrom(stream)
            case DebugUpdate.TAG => DebugUpdate.readFrom(stream)
            case _ => throw new java.io.IOException("Unexpected tag value")
        }
    }
}

File: clients/kotlin/src/main/kotlin/ai_cup_22/codegame/ServerMessage.kt
package ai_cup_22.codegame

import ai_cup_22.util.StreamUtil

/**
 * Message sent from server
 */
abstract class ServerMessage {
    /**
     * Write ServerMessage to output stream
     */
    @Throws(java.io.IOException::class)
    abstract fun writeTo(stream: java.io.OutputStream)

    companion object {
        /**
         * Read ServerMessage from input stream
         */
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): ServerMessage {
            when (StreamUtil.readInt(stream)) {
                UpdateConstants.TAG -> return UpdateConstants.readFrom(stream)
                GetOrder.TAG -> return GetOrder.readFrom(stream)
                Finish.TAG -> return Finish.readFrom(stream)
                DebugUpdate.TAG -> return DebugUpdate.readFrom(stream)
                else -> throw java.io.IOException("Unexpected tag value")
            }
        }
    }

    /**
     * Update constants
     */
    class UpdateConstants : ServerMessage {
        /**
         * New constants
         */
        var constants: ai_cup_22.model.Constants
    
        constructor(constants: ai_cup_22.model.Constants) {
            this.constants = constants
        }
    
        /**
         * Write UpdateConstants to output stream
         */
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
            constants.writeTo(stream)
        }
    
        /**
         * Get string representation of UpdateConstants
         */
        override fun toString(): String {
            var stringBuilder = StringBuilder("UpdateConstants { ")
            stringBuilder.append("constants: ")
            stringBuilder.append(constants)
            stringBuilder.append(" }")
            return stringBuilder.toString()
        }
    
        companion object {
            val TAG = 0
    
            /**
             * Read UpdateConstants from input stream
             */
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): UpdateConstants {
                var constants: ai_cup_22.model.Constants
                constants = ai_cup_22.model.Constants.readFrom(stream)
                return UpdateConstants(constants)
            }
        }
    }

    /**
     * Get order for next tick
     */
    class GetOrder : ServerMessage {
        /**
         * Player's view
         */
        var playerView: ai_cup_22.model.Game
        /**
         * Whether app is running with debug interface available
         */
        var debugAvailable: Boolean
    
        constructor(playerView: ai_cup_22.model.Game, debugAvailable: Boolean) {
            this.playerView = playerView
            this.debugAvailable = debugAvailable
        }
    
        /**
         * Write GetOrder to output stream
         */
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
            playerView.writeTo(stream)
            StreamUtil.writeBoolean(stream, debugAvailable)
        }
    
        /**
         * Get string representation of GetOrder
         */
        override fun toString(): String {
            var stringBuilder = StringBuilder("GetOrder { ")
            stringBuilder.append("playerView: ")
            stringBuilder.append(playerView)
            stringBuilder.append(", ")
            stringBuilder.append("debugAvailable: ")
            stringBuilder.append(debugAvailable)
            stringBuilder.append(" }")
            return stringBuilder.toString()
        }
    
        companion object {
            val TAG = 1
    
            /**
             * Read GetOrder from input stream
             */
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): GetOrder {
                var playerView: ai_cup_22.model.Game
                playerView = ai_cup_22.model.Game.readFrom(stream)
                var debugAvailable: Boolean
                debugAvailable = StreamUtil.readBoolean(stream)
                return GetOrder(playerView, debugAvailable)
            }
        }
    }

    /**
     * Signifies end of the game
     */
    class Finish : ServerMessage {
    
        constructor() {
        }
    
        /**
         * Write Finish to output stream
         */
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
        }
    
        /**
         * Get string representation of Finish
         */
        override fun toString(): String {
            var stringBuilder = StringBuilder("Finish { ")
            stringBuilder.append(" }")
            return stringBuilder.toString()
        }
    
        companion object {
            val TAG = 2
    
            /**
             * Read Finish from input stream
             */
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): Finish {
                return Finish()
            }
        }
    }

    /**
     * Debug update
     */
    class DebugUpdate : ServerMessage {
        /**
         * Displayed tick
         */
        var displayedTick: Int
    
        constructor(displayedTick: Int) {
            this.displayedTick = displayedTick
        }
    
        /**
         * Write DebugUpdate to output stream
         */
        @Throws(java.io.IOException::class)
        override fun writeTo(stream: java.io.OutputStream) {
            StreamUtil.writeInt(stream, TAG)
            StreamUtil.writeInt(stream, displayedTick)
        }
    
        /**
         * Get string representation of DebugUpdate
         */
        override fun toString(): String {
            var stringBuilder = StringBuilder("DebugUpdate { ")
            stringBuilder.append("displayedTick: ")
            stringBuilder.append(displayedTick)
            stringBuilder.append(" }")
            return stringBuilder.toString()
        }
    
        companion object {
            val TAG = 3
    
            /**
             * Read DebugUpdate from input stream
             */
            @Throws(java.io.IOException::class)
            fun readFrom(stream: java.io.InputStream): DebugUpdate {
                var displayedTick: Int
                displayedTick = StreamUtil.readInt(stream)
                return DebugUpdate(displayedTick)
            }
        }
    }
}